server/src-lib/Hasura/Tracing.hs

```diff
@@ -198,13 +201,30 @@ instance MonadTrace m => MonadTrace (ExceptT e m) where
   currentReporter = lift currentReporter
   attachMetadata = lift . attachMetadata
 
+-- | Encode Word64 to 16 character hex string
+word64ToHex :: Word64 -> Text
+word64ToHex randNum = bsToTxt $ Hex.encode numInBytes
+  where numInBytes = BL.toStrict (BIN.encode  randNum)
+
+-- | Decode  16 character hex string to Word64
+-- | TODO: Do we need to check if it has an invalid base16 sequence? 
+-- |       Hex.Decode returns two tuples, first properly decoded data, second - first element where invalid char start
+hexToWord64 :: Text -> Word64
+hexToWord64 randText = BIN.decode decodedHex
+  where decodedHex = BL.fromStrict $ fst $ Hex.decode $ txtToBs randText
+
 -- | Inject the trace context as a set of HTTP headers.
 injectHttpContext :: TraceContext -> [HTTP.Header]
 injectHttpContext TraceContext{..} = 
-  [ ("X-Hasura-TraceId", fromString (show tcCurrentTrace))
-  , ("X-Hasura-SpanId", fromString (show tcCurrentSpan))
-  ]
-  
+  case tcCurrentParent of
+    Nothing -> [ ("X-B3-TraceId", txtToBs $ word64ToHex tcCurrentTrace)
+               , ("X-B3-SpanId", txtToBs $ word64ToHex tcCurrentSpan)
+               ]
+    Just parentID ->  [ ("X-B3-TraceId", txtToBs $ word64ToHex tcCurrentTrace)
+                      , ("X-B3-SpanId", txtToBs $ word64ToHex tcCurrentSpan)
+                      , ("X-B3-ParentSpanId", txtToBs $ word64ToHex parentID)
```

Comment:
I'm not sure what the difference between Span and ParentSpan is supposed to be in B3, and why it requires both. Do you know?

Reply:
So the Zipkin docs say

> This is an optional ID that will only be present on child spans. That is the span without a parent id is considered the root of the trace.

So I assume this means it's useful just to know if this is a root span? I guess it doesn't hurt to include it.

Reply:
Yup! That's totally right, It helps us to differentiate between root span and child span i.e when there is no `ParentSpanID` it means that it is the root of the trace tree. I think it's okay to have it since that's how the propagation is described in format.